Implement override keyword (#7563)
ethereum/solidity: docs/contracts/inheritance.rst

is called, except when the contract name is explicitly given or the
``super`` keyword is used.

> All functions overriding a base function must specify the ``override`` keyword.

Reviewer: Please link the other section here ("for details see ...")

    }

    contract Base1 is mortal {
        function kill() public override { /* do cleanup 1 */ super.kill(); }
    }

    contract Base2 is mortal {
        function kill() public override { /* do cleanup 2 */ super.kill(); }
    }

    contract Final is Base1, Base2 {

>         function kill() public override(Base1, Base2) { Base2.kill(); }

Reviewer: Maybe better to also use `super.kill()` here.

although its type is known. This is similar for ordinary
virtual method lookup.

.. _function-overriding:

.. index:: ! overriding;function

Function Overriding
===================

Base functions can be overridden by inheriting contracts to change their
behavior. The overriding function must then use the ``override`` keyword in the
function header as shown in this example:

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base
    {
        function foo() public {}
    }

    contract Middle is Base {}

    contract Inherited is Middle
    {
        function foo() public override {}
    }

For multiple inheritances, the most derived base contracts that defines the same
function must be specified explicitly after the ``override`` keyword.
Additionally, if a contract inherits the same function from multiple (unrelated)

> bases, it has to explicitly override it:

Reviewer: ```suggestion
For multiple inheritance, the most derived base contracts that define the same
function must be specified explicitly after the ``override`` keyword.
In other words, you have to specify all base contracts that define the same function and have not yet been overridden by another base contract (on some path through the inheritance graph).
Additionally, if a contract inherits the same function from multiple (unrelated)
bases, it has to explicitly override it:
```

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base1
    {
        function foo() public {}
    }

    contract Base2
    {
        function foo() public {}
    }

    contract Inherited is Base1, Base2
    {
        // Derives from multiple bases defining foo(), so we must explicitly
        // override it
        function foo() public override(Base1, Base2) {}
    }

> A function defined in a common base contract must not be explicitly overridden

Reviewer: ```suggestion
A function defined in a common base contract does not have to be explicitly overridden
```

when used with multiple inheritance:

::

    pragma solidity >=0.5.0 <0.7.0;

    contract A { function f() public pure{} }
    contract B is A {}
    contract C is A {}
    // No explicit override required
    contract D is B, C {}

.. _modifier-overriding:

.. index:: ! overriding;modifier

Modifier Overriding
===================

> Function modifiers can override each other. This works in a similar way to function

Reviewer: Either "same way" or "same way (up to the fact that there is no overloading for modifiers)"

although its type is known. This is similar for ordinary
virtual method lookup.

.. _function-overriding:

.. index:: ! overriding;function

Function Overriding
===================

Base functions can be overridden by inheriting contracts to change their
behavior. The overriding function must then use the ``override`` keyword in the
function header as shown in this example:

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base
    {
        function foo() public {}
    }

    contract Middle is Base {}

    contract Inherited is Middle
    {
        function foo() public override {}
    }

> For multiple inheritance, the most derived base contracts that defines the same

Reviewer 2: I feel like this should be

```suggestion
For multiple inheritances, the most derived base contracts that defines the same
```

> function must be specified explicitly. Additionally, the contract deriving from multiple bases

Reviewer: ```suggestion
function must be specified explicitly after the ``override`` keyword. Additionally, if a contract inherits the same function from multiple (unrelated) bases, it has to explicitly override it:
```

defining the same function must explicitly override those functions:

::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base1
    {
        function foo() public {}
    }

    contract Base2
    {
        function foo() public {}
    }

    contract Inherited is Base1, Base2
    {
        // Derives from multiple bases defining foo(), so we must explicitly
        // override it
        function foo() public override(Base1, Base2) {}
    }

> 

Reviewer: Can you also add an example (and a test if we do not have it yet) like this:
```
contract A { function f() public pure{} }
contract B is A {}
contract C is A {}
contract D is B, C {}
```

The function that D inherits via B and C is exactly the same function `f` defined in the same base and thus it does not have to be explicitly overridden.

.. _constructor:

> Function Overriding

Reviewer 3: I think it would be also good to add some documentation on modifier inheritance here https://solidity.readthedocs.io/en/v0.5.12/structure-of-a-contract.html#function-modifiers.

===================

> Base functions can be overridden by inheriting contracts to change their

Reviewer: The `virtual` change will be done in a different PR?

Author: Yes

> behavior. The overriding function must then use the ``override`` keyword in the

Reviewer: What is the currently implemented logic with regards to 3-layer inheritance where the middle contract does not mention the function?

Author: THat is in this implementation allowed and working

Reviewer: Then please explain it here.

Author: Do you consider that behavior uncommon? I figured that's the expected behavior and only if it behaves differently it would be worth documenting..

Reviewer: Below, you use the word "direct base", so I assume when you use "base" here, you mean "direct or indirect base". This makes me conclude that it is allowed to leave out function definitions in the middle contract. If that is so, I would still like this to be stated explicitly because it will reduce a big amount of confusion.

Author: my intention was that one should use the most derived base. I'll try to rephrase it again and I'll add this point, too

function header as shown in this example::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base
    {
        function foo() public {}
    }

    contract Inherited is Base
    {
        function foo() public override {}
    }

> For multiple inheritance, all direct base contracts must be specified

Reviewer: ```suggestion
For multiple inheritance, all direct base contracts that define the same function must be specified
```

explicitly::

    pragma solidity >=0.5.0 <0.7.0;

    contract Base1
    {
        function foo() public {}
    }

    contract Base2
    {
        function foo() public {}
    }

    contract Inherited is Base1, Base2
    {
        function foo() public override(Base1, Base2) {}
    }

>    .. index:: ! overriding;function

Reviewer: Doesn't the index entry need to be mentioned before the section header?

Author: I tried imitating what I saw, not sure how it works

Author: I think it's fine now
